daemon/api_sideload_n_try_test.go

```diff
@@ -508,6 +509,142 @@ func (s *sideloadSuite) TestInstallPathSystemRestartImmediate(c *check.C) {
+}
+
+func (s *sideloadSuite) TestSideloadUsePreciselyAllMemory(c *check.C) {
+	s.daemonWithOverlordMockAndStore()
+
+	buf := make([]byte, daemon.MaxReadBuflen)
+	_, err := rand.Read(buf)
+	c.Assert(err, check.IsNil)
+
+	body := "----hello--\r\n" +
+		"Content-Disposition: form-data; name=\"devmode\"\r\n" +
+		"\r\n" +
+		string(buf) +
+		"\r\n" +
+		"----hello--\r\n"
+
+	req, err := http.NewRequest("POST", "/v2/snaps", bytes.NewBufferString(body))
+	c.Assert(err, check.IsNil)
+	req.Header.Set("Content-Type", "multipart/thing; boundary=--hello--")
+
+	// using the maximum memory doesn't cause the failure (not having a snap file does)
+	apiErr := s.errorReq(c, req, nil)
+	c.Check(apiErr.Message, check.Equals, `cannot find "snap" file field in provided multipart/form-data payload`)
+}
+
+func (s *sideloadSuite) TestCleanUpTempFilesIfRequestFailed(c *check.C) {
+	s.daemonWithOverlordMockAndStore()
+
+	// write file parts
+	body := "----hello--\r\n"
+	for _, name := range []string{"one", "two"} {
+		body += fmt.Sprintf(
+			"Content-Disposition: form-data; name=\"snap\"; filename=\"%s\"\r\n"+
+				"\r\n"+
+				"xyzzy\r\n", name)
+	}
+
+	// make the request fail
+	buf := make([]byte, daemon.MaxReadBuflen+1)
+	_, err := rand.Read(buf)
+	c.Assert(err, check.IsNil)
+
+	body += "----hello--\r\n" +
+		"Content-Disposition: form-data; name=\"devmode\"\r\n" +
+		"\r\n" +
+		string(buf) +
+		"\r\n" +
+		"----hello--\r\n"
+
+	req, err := http.NewRequest("POST", "/v2/snaps", bytes.NewBufferString(body))
+	c.Assert(err, check.IsNil)
+	req.Header.Set("Content-Type", "multipart/thing; boundary=--hello--")
+
+	apiErr := s.errorReq(c, req, nil)
+	c.Check(apiErr, check.NotNil)
+	matches, err := filepath.Glob(filepath.Join(dirs.SnapBlobDir, dirs.LocalInstallBlobTempPrefix+"*"))
+	c.Assert(err, check.IsNil)
+	c.Check(matches, check.HasLen, 0)
+}
+
+func (s *sideloadSuite) TestCleanUpUnusedTempSnapFiles(c *check.C) {
+	body := "----hello--\r\n" +
+		"Content-Disposition: form-data; name=\"devmode\"\r\n" +
+		"\r\n" +
+		"true\r\n" +
+		"----hello--\r\n" +
+		"Content-Disposition: form-data; name=\"snap\"; filename=\"one\"\r\n" +
+		"\r\n" +
+		"xyzzy\r\n" +
+		"----hello--\r\n" +
+		"Content-Disposition: form-data; name=\"snap\"; filename=\"two\"\r\n" +
+		"\r\n" +
+		// sideloadCheck checks that the snap file passed to the change has contents "xyzzy" so
+		// having a different body here tests that the second file isn't passed to change
+		"bla\r\n" +
+		"----hello--\r\n"
+
+	head := map[string]string{"Content-Type": "multipart/thing; boundary=--hello--"}
+	chgSummary, _ := s.sideloadCheck(c, body, head, "local", snapstate.Flags{RemoveSnapPath: true, DevMode: true})
+	c.Check(chgSummary, check.Equals, `Install "local" snap from file "one"`)
+
+	matches, err := filepath.Glob(filepath.Join(dirs.SnapBlobDir, dirs.LocalInstallBlobTempPrefix+"*"))
+	c.Assert(err, check.IsNil)
+	// only the file passed into the change remains
+	c.Check(matches, check.HasLen, 1)
```

Comment:
```suggestion
	// only the file passed into the change remains, which is the first one
	// passed in the request
	c.Check(matches, check.HasLen, 1)
	c.Check(matches[0], testutil.FileEquals, "xyzzy")
```

Reply:
I think sideloadCheck already does this check